### PR TITLE
printing - fixes issue 21814 and now some expressions are printed correctly

### LINKED ISSUE #21814: Printing of unevaluated Mul needs brackets
The second version here should use brackets or something:
```python
In [16]: Mul(Rational(1, 2), x + y, evaluate=False)
Out[16]: 
x + y
─────
  2  

In [17]: Mul(x + y, Rational(1, 2), evaluate=False)
Out[17]: x + y⋅1/2
```

### PR DESCRIPTION
**Pull request by praneethratna**
#### Brief description of what is fixed or changed
Fixes #21814 and the now the Mul expression Mul(x + y, Rational(1, 2), evaluate=False) is printed as (x+y)*2 also fixed the printing in pretty and upretty.

#### Release Notes

* printing
     * Fixed printing of unevaluated Mul which now outputs with brackets

### CODE REVIEW

**Comment by praneethratna** [comment]
@oscarbenjamin pleace review

**Comment by namannimmo10** [comment]
Looks like LaTeX printing is also broken:

```python
>>> latex(Mul(x + y, Rational(1, 2), evaluate=False))
'\\left(x + y\\right) \\frac{1}{2}'
>>> latex(Mul(Rational(1, 2), x + y, evaluate=False))
'\\frac{x + y}{2}'
>>>
```

**Comment by praneethratna** [comment]
> Looks like LaTeX printing is also broken:
> 
> ```python
> >>> latex(Mul(x + y, Rational(1, 2), evaluate=False))
> '\\left(x + y\\right) \\frac{1}{2}'
> >>> latex(Mul(Rational(1, 2), x + y, evaluate=False))
> '\\frac{x + y}{2}'
> >>>
> ```

@namannimmo10 now both latex(Mul(x + y, Rational(1, 2), evaluate=False)) and latex(Mul(Rational(1, 2), x + y, evaluate=False)) outputs '\\frac{x + y}{2}' well i guess it is solved then?

**Comment by namannimmo10** [comment]
@oscarbenjamin, I just had a look at https://github.com/sympy/sympy/pull/19450. If we want to show the identities explicitly when `evaluate=False`, then is this the desired behavior:
```python
In [5]: Mul(S.One, x + y, evaluate=False)
Out[5]: 1⋅(x + y)

In [6]: Mul(x - y, S.One, evaluate=False)
Out[6]: x - y

In [7]: Mul(Rational(1, 2), x - y, S.One, x + y, evaluate=False)
Out[7]: 1/2⋅(x - y)⋅1⋅(x + y)

In [8]: Mul(x + y, Rational(3, 4), S.One, y - z, evaluate=False)
Out[8]: 
3⋅(x + y)⋅(y - z)
─────────────────
        4        

In [9]: Mul(x + y, Rational(1, 1), Rational(3, 4), Rational(5, 6),evaluate=False)
Out[9]: 
3⋅5⋅(x + y)
───────────
    4⋅6    

In [10]: Mul(Rational(1, 1), x + y, Rational(3, 4), Rational(5, 6),evaluate=False)
Out[10]: 1⋅(x + y)⋅3/4⋅5/6

In [12]: Mul(Rational(3, 4), x + y, S.One, y - z, evaluate=False)
Out[12]: 3/4⋅(x + y)⋅1⋅(y - z)
```
Why does it print `1` explicitly in `In [12]` but not in `In [8]`?

**Comment by oscarbenjamin** [comment]
> Why does it print `1` explicitly in `In [12]` but not in `In [8]`?

It should show the `1` in `[8]` as well but I don't know what commit you are testing. Current master gives:
```python
In [2]: Mul(x + y, Rational(3, 4), S.One, y - z, evaluate=False)
Out[2]: x + y⋅3/4⋅1⋅(y - z)
```
That's correct except that brackets are needed for `(x + y)`.

**Comment by praneethratna** [comment]
> > Why does it print `1` explicitly in `In [12]` but not in `In [8]`?
> 
> It should show the `1` in `[8]` as well but I don't know what commit you are testing. Current master gives:
> 
> ```python
> In [2]: Mul(x + y, Rational(3, 4), S.One, y - z, evaluate=False)
> Out[2]: x + y⋅3/4⋅1⋅(y - z)
> ```
> 
> That's correct except that brackets are needed for `(x + y)`.

@oscarbenjamin This branch fixes that , On this branch Mul(x + y, Rational(3, 4), S.One, y - z, evaluate=False)
outputs (x + y)*(3/4)*1*(y - z)

File: sympy/printing/pretty/tests/test_pretty.py [hunk 1]
```diff
@@ -7561,3 +7561,7 @@ def test_diffgeom():
     assert pretty(rect) == "rect"
     b = BaseScalarField(rect, 0)
     assert pretty(b) == "x"
+
+def test_issue_21814():
+    assert upretty(Mul(x + y, Rational(1, 2), evaluate=False)) == 'x + y\n─────\n  2  '
+    assert upretty(Mul(Rational(1, 2), x + y, evaluate=False)) == 'x + y\n─────\n  2  '
```
Comment (oscarbenjamin) [inline sympy/printing/pretty/tests/test_pretty.py:7567]: These tests could go with the others:
https://github.com/sympy/sympy/blob/4ab830d193a9d2c197d4e95ee031ca0865d011e0/sympy/printing/pretty/tests/test_pretty.py#L954-L998

File: sympy/printing/tests/test_latex.py [hunk 2]
```diff
@@ -2755,3 +2755,7 @@ def test_pickleable():
 def test_printing_latex_array_expressions():
     assert latex(ArraySymbol("A", 2, 3, 4)) == "A"
     assert latex(ArrayElement("A", (2, 1/(1-x), 0))) == "{{A}_{2, \\frac{1}{1 - x}, 0}}"
+
+def test_issue_21814():
+    assert latex(Mul(x + y, Rational(1, 2), evaluate=False)) == "\\frac{x + y}{2}"
+    assert latex(Mul(Rational(1, 2), x + y, evaluate=False)) == "\\frac{x + y}{2}"
```
Comment (oscarbenjamin) [inline sympy/printing/tests/test_latex.py:2761]: These could go with the related tests:
https://github.com/sympy/sympy/blob/1d9fb1edeec2972dbc472f24064128e16bf39640/sympy/printing/tests/test_latex.py#L91-L108

**Comment by oscarbenjamin** [comment]
The examples @namannimmo10 listed above could go in the tests if they aren't already.

**Comment by praneethratna** [comment]
> The examples @namannimmo10 listed above could go in the tests if they aren't already.

@oscarbenjamin where should the tests for these be added ? in both latex and pretty tests?

File: sympy/printing/tests/test_latex.py (hunk 2 above)
Comment (praneethratna) [reply sympy/printing/tests/test_latex.py]: @oscarbenjamin made the changes!

**Comment by praneethratna** [comment]
@oscarbenjamin can this be merged?

File: sympy/printing/latex.py [hunk 3]
```diff
@@ -546,7 +546,7 @@ def convert_args(args):
         # XXX: _print_Pow calls this routine with instances of Pow...
         if isinstance(expr, Mul):
             args = expr.args
-            if args[0] is S.One or any(isinstance(arg, Number) for arg in args[1:]):
+            if args[0] is S.One or any(isinstance(arg, Number) for arg in args[1:]) and args[0].is_number:
```
Comment (oscarbenjamin) [inline sympy/printing/latex.py:549]: Can you explain what this change is for?

I don't think `is_number` is the right thing to check here.

**Comment by oscarbenjamin** [comment]
> The examples @namannimmo10 listed above could go in the tests if they aren't already.

Are these examples already tested?

File: sympy/printing/latex.py (hunk 3 above)
Comment (praneethratna) [reply sympy/printing/latex.py:549]: I added that to make sure latex(Mul(x + y, Rational(1, 2), evaluate=False)) outputs '\\frac{x + y}{2}' which before is returning a different output '\\left(x + y\\right) \\frac{1}{2}' due to checking for second arg as a number.

**Comment by praneethratna** [comment]
> > The examples @namannimmo10 listed above could go in the tests if they aren't already.
> 
> Are these examples already tested?

I'm confused on to where these examples should be tested , in the latex ones or pretty one?

**Comment by oscarbenjamin** [comment]
> I'm confused on to where these examples should be tested , in the latex ones or pretty one?

They could be tested for both.

**Pull request closed without merging**